### PR TITLE
gh-91053: Add an optional callback that is invoked whenever a function is modified

### PR DESCRIPTION
**Pull request by mpage**
JIT compilers may need to invalidate compiled code when a function is modified (e.g. if its code object is modified). This adds the ability to set a callback that, when set, is called each time a function is modified.

* Issue: gh-91053

### CODE REVIEW

**Comment by mpage** [comment]
The failing test appears unrelated to this diff. Is there a way that I can re-run the test without adding additional commits?

**Comment by erlend-aasland** [comment]
> The failing test appears unrelated to this diff. Is there a way that I can re-run the test without adding additional commits?

You can merge in `main`

**Comment by zooba** [comment]
> You can merge in `main`

Or any core dev (and probably triager) can trigger a rerun, which I did already and it's now passed.

File: Include/cpython/funcobject.h
Comment (erlend-aasland) [inline Include/cpython/funcobject.h]: The other added constants use the imperative mood. Suggesting to be consistent and also use the imperative mood here.

File: Include/cpython/funcobject.h
Comment (erlend-aasland) [inline Include/cpython/funcobject.h:144]: IMO, it unneccesary messy to use two macros to create a list of five enums; I'd simply spell them out. I guess it is fine to keep it this way also, but at least `#undef FOREACH_FUNC_EVENT` when you're done with it.

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: PEP7-nit:

```suggestion
static PyObject *
get_id(PyObject *obj)
{
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: Please add a comment that this is a borrowed reference.

```suggestion
    PyObject *builtins = PyEval_GetBuiltins();  // borrowed ref.
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: PEP 7-nit:

```suggestion
call_pyfunc_watcher(PyObject *watcher, PyFunction_WatchEvent event,
                    PyFunctionObject *func, PyObject *new_value)
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: PEP7

```suggestion
    }
    else {
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: PEP7: wrap at 79 chars

```suggestion
first_watcher_callback(PyFunction_WatchEvent event, PyFunctionObject *func,
                       PyObject *new_value)
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: PEP7: wrap at 79 chars

```suggestion
second_watcher_callback(PyFunction_WatchEvent event, PyFunctionObject *func,
                        PyObject *new_value)
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: PEP7: 4 space indent

```suggestion
    first_watcher_callback,
    second_watcher_callback,
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: Shortcut:

```suggestion
    pyfunc_watchers[idx] = Py_NewRef(func);
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: PEP7: 4 spaces indent

```suggestion
        if (watcher_ids[i] == wid) {
            idx = i;
            break;
        }
```

File: Objects/funcobject.c
Comment (erlend-aasland) [inline Objects/funcobject.c]: Convention (at least in this file) is to start the exception message with a lower-case letter. Also, wrap at 79 chars.

```suggestion
        PyErr_Format(PyExc_ValueError, "invalid func watcher ID %d",
                     watcher_id);
```

File: Objects/funcobject.c
Comment (erlend-aasland) [inline Objects/funcobject.c]:
```suggestion
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: Watcher id's less than 0 are not allowed (unless I misread something), so the lower bounds check should be `watcher_id < 0`, right?

Comment (erlend-aasland) [reply Modules/_testcapi/func_events.c]: Also, the upper bound should be `FUNC_MAX_WATCHERS` inclusive.

File: Objects/funcobject.c
Comment (erlend-aasland) [inline Objects/funcobject.c]: Ditto

```suggestion
        PyErr_Format(PyExc_ValueError, "no func watcher set for ID %d",
                     watcher_id);
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: PEP 7

```suggestion
_PyTestCapi_Init_FuncEvents(PyObject *m)
{
```

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: FYI, you can get rid of the whole `add_event` helper if you use `PyModule_AddIntMacro`.

File: Doc/c-api/function.rst
Comment (erlend-aasland) [inline Doc/c-api/function.rst]: Nit: semantic line-breaks will make this paragraph more neatly formatted. Personal preference only; you may ignore this suggestion. Also, consistently use imperative mood.

```suggestion
   Register *callback* as a function watcher for the current interpreter.
   Return an id which may be passed to :c:func:`PyFunction_ClearWatcher`.
   In case of error (e.g. no more watcher IDs available),
   return ``-1`` and set an exception.
```

Comment (erlend-aasland) [reply Doc/c-api/function.rst]: Also, throughout the PR you use both lower-case `id` and upper-case `ID` (docs, comments, exception messages, etc.). I'd consider sticking to one of them for consistency.

File: Doc/c-api/function.rst
Comment (erlend-aasland) [inline Doc/c-api/function.rst]: I'd append _"and set an exception"_ for the error description. For example:

```suggestion
   :c:func:`PyFunction_AddWatcher` for the current interpreter.
   Return ``0`` on success, or ``-1`` and set an exception on error
   (e.g.  if the given *watcher_id* was never registered.)
```

File: Doc/c-api/function.rst
Comment (erlend-aasland) [inline Doc/c-api/function.rst]: Consider formatting these as a list.

File: Doc/c-api/function.rst
Comment (erlend-aasland) [inline Doc/c-api/function.rst]: Consider adding a reference to ```:term:`borrowed reference` ```.

File: Objects/funcobject.c
Comment (erlend-aasland) [inline Objects/funcobject.c:46]: This branch is not tested.

**Review by erlend-aasland — COMMENTED** [review]
Looks good. I added a bunch of PEP 7 comments because we should (almost) always strive to follow the style guide, both for new code and (in most cases) when modifying existing code. I also added some comments regarding docs, inconsistent naming of the constants, and some C API suggestions.

Questions:
- `PyFunction_SetDefaults` should emit a `PyFunction_EVENT_MODIFY_DEFAULTS` event, no?
- `PyFunction_SetKwDefaults` should emit a `PyFunction_EVENT_MODIFY_KWDEFAULTS` event, no?
- Why no events for `func_closure` and `func_annotations` modifications?

Also, please add the Python test code to `Lib/test/test_capi.py` instead of adding a new `Lib/test/test_func_events.py`, like the dict watchers did.

File: Include/cpython/funcobject.h
Comment (mpage) [reply Include/cpython/funcobject.h:144]: This is intentionally left defined to provide others a means to programmatically apply a macro to each function event at compile time (https://en.wikipedia.org/wiki/X_Macro). This is used in `func_events.c` to expose string constants for each function event.

File: Modules/_testcapi/func_events.c
Comment (mpage) [reply Modules/_testcapi/func_events.c]: This is to catch numeric errors when casting from types of potentially different sizes. The bounds checks on watcher ids happen in `PyFunction_ClearWatcher`.

File: Modules/_testcapi/func_events.c
Comment (mpage) [reply Modules/_testcapi/func_events.c]: I ended up keeping things as-is so that I can expose the constants to Python as `SCREAMING_SNAKE_CASE` rather than a mix of that and `PascalCase`.

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [inline Modules/_testcapi/func_events.c]: PEP 7 nit

```suggestion
static PyObject *
get_id(PyObject *obj)
```

**Review by erlend-aasland — COMMENTED** [review]
Thanks for making the changes! I've got one last minor style nit. Also please merge in main and resolve the conflicts.

File: Modules/_testcapi/func_events.c
Comment (mpage) [reply Modules/_testcapi/func_events.c]: Do you use an automatic formatter like `clang-format`? If so, would you mind sharing your config?

**Comment by markshannon** [comment]
I don't like having watchers for function creation and destruction. 

Having those watchers adds overhead for every creation of a comprehension or nested function, which could be an issue for performance for two reasons:
1. The overhead. While the overhead is hard to measure now, that is more a comment on how inefficient function object creation currently is, rather than the overhead being negligible.
2. It prevents the removal of functions by the optimizer. For example, there is no reason for a function to be created for a list comprehension. An optimizer could avoid creating the function and preserve semantics. If creation of a function is always observable, this optimization becomes impossible.

File: Modules/_testcapi/func_events.c
Comment (erlend-aasland) [reply Modules/_testcapi/func_events.c]: Sorry, I don't use an automatic formatter for CPython; I've got PEP 7 in my fingers :) I know there's been discussions about having a PEP 7 check in the CI or doing some kind of automatic formatting, but those ideas have been rejected every time.

See [this Discourse topic](https://discuss.python.org/t/pep-proposal-automatically-formatting-the-cpython-code/5603) for one of the most recent discussions.

Comment (erlend-aasland) [reply Modules/_testcapi/func_events.c]: Paul Ganssle has a `clang-format` config that approximate PEP 7: https://gist.github.com/pganssle/0e3a5f828b4d07d79447f6ced8e7e4db

File: Objects/funcobject.c
Comment (markshannon) [inline Objects/funcobject.c:15]: If we add a bit vector we can skip the iteration when there are no watchers.
Maybe put that test in an inline function and call this function only if there are watchers?

**Comment by carljm** [comment]
Discussed offline with @markshannon and reached these conclusions (Mark please let me know if any of this doesn't match your understanding):

1. In the long term it might be better if instead of setting function vectorcall (and thus needing func-created hook so we can set it), we could rely on the core runtime to tell us when a function is hot and defer to our execution in that case. (This can be made compatible with pre-fork compilation if we eagerly compile code objects when they are created, and then just pull the pointer to the compiled code out of our cache when the runtime says "this function is hot, please execute it for me.") But this is all hypothetical at the moment and still requires a lot of design, implementation, and verification that it actually works, so it's a bit premature to rely on this solution.
2. The "observability of function creation" issue doesn't need to be a blocker here. The documentation for the hook can be clear that optimized execution of Python code is free to elide function-object creation where possible, and if function object creation is optimized away, no function-created hook will be fired, and this is not considered a semantic change.
3. The overhead here really isn't much at all compared to the existing work of creating a function object. Let's try to optimize it a bit more so that in the no-watchers case it's just a single read/test (per Mark's inline comment), and that'll be good enough.

We also discussed a possible alternative design where code objects have a vectorcall pointer and function objects just copy it from the code object; then we wouldn't need func-created hook. But this also adds an extra read to function creation, as well as making code objects bigger, so it's not clear this is really a better approach.

File: Doc/c-api/function.rst
Comment (carljm) [inline Doc/c-api/function.rst:175]: I guess we should add a note here that optimized execution of Python code is permitted to eliminate creation of a function object if possible, and this is not considered a semantic change, even though it is observable if a function watcher is registered.

File: Modules/_testcapi/parts.h
Comment (erlend-aasland) [inline Modules/_testcapi/parts.h]: FYI, I recently split out the dict and type watcher tests into `Modules/_testcapi/watchers.c` and `Lib/test/test_capi/test_watchers.py`. Would you mind redirecting your test code to those files? Sorry for the inconvenience.

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
LGTM, thanks! I'll wait for @markshannon's thumbs up before merging.

**Review by markshannon — APPROVED** [review]

**Pull request merged**
